--- a/BleKeyboard.cpp
+++ b/BleKeyboard.cpp
@@ -1,3 +1,5 @@
+#include "BleKeyboard.h"
+
 #if defined(CONFIG_ARDUHAL_ESP_LOG)
   #include "esp32-hal-log.h"
   #define LOG_TAG ""
@@ -14,7 +16,6 @@
   #define DPRINTF(fmt, ...)   do {} while(0)
 #endif // CONFIG_ARDUHAL_ESP_LOG
 
-#include "BleKeyboard.h"
 
 #if defined(USE_NIMBLE)
 
@@ -102,7 +103,7 @@ void BleKeyboard::begin(void)
   configuration.setAutoReport(false);
   configuration.setControlPacket(CONTROL_PACKET_SIZE);
 
-  BLEDevice::init(deviceName);
+  BLEDevice::init(String(deviceName.c_str()));
   BLEServer *pServer = BLEDevice::createServer();
   pServer->setCallbacks(this);
 
@@ -113,13 +114,18 @@ void BleKeyboard::begin(void)
   hid->reportMap((uint8_t*)descriptor, sizeof(descriptor));
   hid->startServices();
 
-  hid->manufacturer()->setValue(deviceManufacturer);
+  hid->manufacturer()->setValue(String(deviceManufacturer.c_str()));
 
-  hid->pnp(0x02, 0xe502, 0xa111, 0x0210);
+  hid->pnp(0x02, vid, pid, version);
   hid->hidInfo(0x00,0x01);
 
+#if defined(USE_NIMBLE)
+
+  BLEDevice::setSecurityAuth(true, true, true);
+
+#else
+
   BLESecurity *pSecurity = new BLESecurity();
-  pSecurity->setAuthenticationMode(ESP_LE_AUTH_BOND);
+  pSecurity->setAuthenticationMode(ESP_LE_AUTH_REQ_SC_MITM_BOND);
 
+#endif // USE_NIMBLE
   
@@ -164,6 +170,17 @@ void BleKeyboard::end(void)
 {
 }
 
+void BleKeyboard::set_vendor_id(uint16_t vid) {
+	this->vid = vid;
+}
+
+void BleKeyboard::set_product_id(uint16_t pid) {
+	this->pid = pid;
+}
+
+void BleKeyboard::set_version(uint16_t version) {
+	this->version = version;
+}
+
 bool BleKeyboard::isConnected(void) {
   return this->connected;
 }
@@ -479,11 +496,24 @@ void BleKeyboard::onConnect(BLEServer* pServer) {
   this->connected = true;
 
   BLE2902* desc = (BLE2902*)this->inputKeyboard->getDescriptorByUUID(BLEUUID((uint16_t)0x2902));
+
+#if !defined(USE_NIMBLE)
+
+  BLE2902* desc = (BLE2902*)this->inputKeyboard->getDescriptorByUUID(BLEUUID((uint16_t)0x2902));
+  desc->setNotifications(true);
+  desc = (BLE2902*)this->inputMediaKeys->getDescriptorByUUID(BLEUUID((uint16_t)0x2902));
+  desc->setNotifications(true);
+
+#endif // !USE_NIMBLE
+
 }
 
 void BleKeyboard::onDisconnect(BLEServer* pServer) {
   this->connected = false;
+
+  BLE2902* desc = (BLE2902*)this->inputKeyboard->getDescriptorByUUID(BLEUUID((uint16_t)0x2902));
+  desc->setNotifications(false);
+  desc = (BLE2902*)this->inputMediaKeys->getDescriptorByUUID(BLEUUID((uint16_t)0x2902));
+  desc->setNotifications(false);
 
-#endif  // !USE_NIMBLE
+
+#endif // !USE_NIMBLE
